modules/tensor_mechanics/src/kernels/lagrangian/TotalLagrangianStressDivergenceBase.C

```diff
@@ -0,0 +1,117 @@
+    _dpk1(getMaterialPropertyByName<RankFourTensor>(_base_name + "pk1_jacobian"))
+{
+}
+
+template <class G>
+RankTwoTensor
+TotalLagrangianStressDivergenceBase<G>::gradTest(unsigned int component)
+{
+  // F-bar doesn't modify the test function
+  return G::gradOp(component, _grad_test[_i][_qp], _test[_i][_qp], _q_point[_qp]);
+}
+
+template <class G>
+RankTwoTensor
+TotalLagrangianStressDivergenceBase<G>::gradTrial(unsigned int component)
+{
+  return _stabilize_strain ? gradTrialStabilized(component) : gradTrialUnstabilized(component);
+}
+
+template <class G>
+RankTwoTensor
+TotalLagrangianStressDivergenceBase<G>::gradTrialUnstabilized(unsigned int component)
+{
+  // Without F-bar stabilization, simply return the gradient of the trial functions
+  return G::gradOp(component, _grad_phi[_j][_qp], _phi[_j][_qp], _q_point[_qp]);
+}
+
+template <class G>
+RankTwoTensor
+TotalLagrangianStressDivergenceBase<G>::gradTrialStabilized(unsigned int component)
+{
+  // The base unstabilized trial function gradient
+  const auto Gb = G::gradOp(component, _grad_phi[_j][_qp], _phi[_j][_qp], _q_point[_qp]);
+  // The average trial function gradient
+  const auto Ga = _avg_grad_trial[component][_j];
+
+  // The F-bar stabilization depends on kinematics
+  if (_large_kinematics)
+  {
+    // Horrible thing, see the documentation for how we get here
+    const Real dratio = std::pow(_F_avg[_qp].det() / _F_ust[_qp].det(), 1.0 / 3.0);
+    const Real fact = (_F_avg[_qp].inverse().transpose().doubleContraction(Ga) -
+                       _F_ust[_qp].inverse().transpose().doubleContraction(Gb)) /
+                      3.0;
+    return dratio * (Gb + fact * _F_ust[_qp]);
+  }
+
+  // The small kinematics modification is linear
+  return Gb + (Ga.trace() - Gb.trace()) / 3.0 * RankTwoTensor::Identity();
+}
+
+template <class G>
+void
+TotalLagrangianStressDivergenceBase<G>::precalculateJacobianDisplacement(unsigned int component)
+{
+  // For total Lagrangian, the averaging is taken on the reference frame regardless of geometric
+  // nonlinearity. Convenient!
+  for (auto j : make_range(_phi.size()))
+    _avg_grad_trial[component][j] = StabilizationUtils::elementAverage(
+        [this, component, j](unsigned int qp)
+        { return G::gradOp(component, _grad_phi[j][qp], _phi[j][qp], _q_point[qp]); },
+        _JxW,
+        _coord);
+}
+
+template <class G>
+Real
+TotalLagrangianStressDivergenceBase<G>::computeQpResidual()
+{
+  return gradTest(_alpha).doubleContraction(_pk1[_qp]);
+}
+
+template <class G>
+Real
+TotalLagrangianStressDivergenceBase<G>::computeQpJacobianDisplacement(unsigned int alpha,
+                                                                      unsigned int beta)
+{
+  // J_{alpha beta} = phi^alpha_{i, J} T_{iJkL} G^beta_{kL}
+  return gradTest(alpha).doubleContraction(_dpk1[_qp] * gradTrial(beta));
+}
+
+template <class G>
+Real
+TotalLagrangianStressDivergenceBase<G>::computeQpJacobianTemperature(unsigned int cvar)
```

Comment:
I checked my derivation and I'm 99% sure this is agnostic to the coordinate system, but could you assuage the remaining 1% of doubt by including the thermal_expansion jactests for the other two coordinate systems?

Reply:
:+1: